lib/fss.ex

```diff
@@ -13,19 +13,102 @@ defmodule FSS do
-              region: String.t(),
               access_key_id: String.t(),
-              secret_access_key: String.t()
+              endpoint: String.t(),
+              region: String.t(),
+              secret_access_key: String.t(),
+              token: String.t() | nil
             }
+
+      def from_system_env() do
+        %__MODULE__{
+          access_key_id: System.get_env("AWS_ACCESS_KEY_ID"),
+          secret_access_key: System.get_env("AWS_SECRET_ACCESS_KEY"),
+          region: System.get_env("AWS_REGION", System.get_env("AWS_DEFAULT_REGION")),
+          token: System.get_env("AWS_SESSION_TOKEN")
+        }
+      end
     end
 
     defmodule Entry do
-      defstruct [:url, :config]
+      defstruct [:bucket, :key, :port, :config]
+
+      @type t :: %__MODULE__{
+              bucket: String.t(),
+              key: String.t(),
+              port: pos_integer(),
+              config: Config.t()
+            }
+
+      def parse(url, opts \\ []) do
+        opts = 
+          opts
+          |> Keyword.validate!(opts, [:config])
+          |> Keyword.put_new_lazy(:config, fn -> FSS.S3.Config.from_system_env() end)
+        uri = URI.parse(url)
+
+        cond do
+          uri.scheme == "s3" ->
+            case uri do
+              %{host: host, path: "/" <> key} when is_binary(host) ->
+                {:ok, %__MODULE__{bucket: host, key: key, port: uri.port, config: opts[:config]}}
+
+              %{host: nil} ->
+                {:error, "host is required"}
+
+              %{path: nil} ->
+                {:error, "path to the resource is required"}
+            end
+
+          uri.scheme in ["http", "https"] ->
+            parts =
+              case String.split(uri.host, ".") do
+                ["s3", region | tail] ->
+                  "/" <> path = uri.path
+                  [bucket, key] = String.split(path, "/", parts: 2)
+                  endpoint = Enum.join(tail, ".")
+
+                  {:ok, [bucket: bucket, key: key, region: region, endpoint: endpoint]}
+
+                [bucket, "s3", region | tail] ->
+                  "/" <> key = uri.path
+                  endpoint = Enum.join(tail, ".")
+
+                  {:ok, [bucket: bucket, key: key, region: region, endpoint: endpoint]}
+
+                _ ->
+                  {:error, "cannot read a valid S3 URI from #{inspect(url)}"}
+              end
+
+            with {:ok, parts} <- parts do
+              config = %FSS.S3.Config{
+                opts[:config]
+                | region: parts[:region],
+                  endpoint: parts[:endpoint]
+              }
+
+              {:ok,
+               %__MODULE__{
+                 bucket: parts[:bucket],
+                 key: parts[:key],
+                 port: uri.port,
+                 config: config
+               }}
+            end
 
-      @type t :: %__MODULE__{url: String.t(), config: Config.t()}
+          true ->
+            {:error, :not_supported}
```

Comment:
```suggestion
            {:error, :invalid_format}
```

Reply:
I added a message for this error. WDYT?
https://github.com/elixir-explorer/explorer/pull/652/files#diff-0cb4ec47f6ec5659387509052e0673b6204edecf2e060b49a141551c6eb7b22bR72